### ISSUE TITLE
enabling postgres extension fails with "already exists"

### ISSUE THREAD

**Issue by jcgruenhage** [ansible-collections/community.postgresql#421]
##### SUMMARY

We have a task that enables a postgres extension on two databases inside of a postgres instance, and sometimes (but not all of the times), it fails on one of them. Looks like this:

```
TASK [Enable 'uuid-ossp' extension on api-component databases] ***************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: DETAIL:  Key (extname)=(uuid-ossp) already exists.
failed: [ci] (item=a-database) => {"ansible_loop_var": "item", "changed": false, "item": "a-database", "msg": "Management of PostgreSQL extension failed: duplicate key value violates unique constraint \"pg_extension_name_index\"\nDETAIL:  Key (extname)=(uuid-ossp) already exists.\n"}
changed: [ci] => (item=another-database)
```

The task for this looks like this:
```yaml
    - name: Enable 'uuid-ossp' extension on api-component databases
      community.postgresql.postgresql_ext:
        name: uuid-ossp
        db: "{{ item }}"
        login_host: "{{ postgres_unix_socket_path }}"
      loop:
        - "{{ a_database_name }}"
        - "{{ another_database_name }}"
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
community.postgresql.postgresql_ext

##### ANSIBLE VERSION
<del>Paste verbatim output from "ansible --version" between quotes</del> Runs in CI, so that's a bit difficult, but it's a debian on which we've installed ansible==6.4.0 and ansible-core==2.13.4. If needed, we can include `ansible --version` in our CI scripts though.

I've skipped a bunch of things that the template requests, because those are fairly redundant with what has already been written above or they are things that aren't really available.

Either way: This is a fairly annoying bug, we've not been able to isolate when it happens yet, we've not seen it on any local deployments, just in CI, and in CI also consistently only for one component. I have no idea what could cause this, but wanted to submit this upstream anyway, to check in whether anyone else has seen this before and also to ask whether anyone would have a hint on what might be causing this?

**Comment by Andersson007** [comment]
@jcgruenhage hi, thanks for reporting the issue!
I don't think anyone has reported anything similar so far.
Is it possible to get output of the task run with the `-vvv` argument? If possible, could you please put it in a comment?

**Comment by jcgruenhage** [comment]
I've enabled `-vvv` on the CI job and added `ansible --version` to our CI scripts, but the heisenbug hasn't shown since. Will reach back out when we have a failure with verbose logging.

**Comment by Andersson007** [comment]
@jcgruenhage OK, thanks, let's close the issue for a while. When you catch the error, please just re-open it.

**Comment by jcgruenhage** [comment]
We've got a failure with `-vvv`!

```
TASK [Enable 'uuid-ossp' extension on workflow-engine databases] ***************
task path: /home/gitlab-runner/builds/famedly/infra/environments/local/playbooks/workflow-engine.yml:59
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_community.postgresql.postgresql_ext_payload_oqzy70qu/ansible_community.postgresql.postgresql_ext_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_ext.py", line 455, in main
  File "/tmp/ansible_community.postgresql.postgresql_ext_payload_oqzy70qu/ansible_community.postgresql.postgresql_ext_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_ext.py", line 275, in ext_create
  File "/usr/local/lib/python3.9/dist-packages/psycopg2/extras.py", line 146, in execute
    return super(DictCursor, self).execute(query, vars)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "pg_extension_name_index"
DETAIL:  Key (extname)=(uuid-ossp) already exists.

failed: [ci.famedly.local] (item=workflow-engine) => {
    "ansible_loop_var": "item",
    "changed": false,
    "invocation": {
        "module_args": {
            "ca_cert": null,
            "cascade": false,
            "db": "workflow-engine",
            "ext": "uuid-ossp",
            "login_host": "/opt/postgresql/sockets",
            "login_password": "",
            "login_unix_socket": "",
            "login_user": "postgres",
            "name": "uuid-ossp",
            "port": 5432,
            "schema": null,
            "session_role": null,
            "ssl_mode": "prefer",
            "state": "present",
            "trust_input": true,
            "version": null
        }
    },
    "item": "workflow-engine",
    "msg": "Management of PostgreSQL extension failed: duplicate key value violates unique constraint \"pg_extension_name_index\"\nDETAIL:  Key (extname)=(uuid-ossp) already exists.\n"
}
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> ESTABLISH LOCAL CONNECTION FOR USER: gitlab-runner
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> EXEC /bin/sh -c 'echo ~gitlab-runner && sleep 0'
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/gitlab-runner/.ansible/tmp `"&& mkdir "` echo /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905 `" && echo ansible-tmp-1678973886.7891605-17637-250160229280905="` echo /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905 `" ) && sleep 0'
Using module file /home/gitlab-runner/builds/famedly/infra/environments/local/lib/python3.9/site-packages/ansible_collections/community/postgresql/plugins/modules/postgresql_ext.py
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> PUT /home/gitlab-runner/.ansible/tmp/ansible-local-1367l34s3xs1/tmplkg1dq6g TO /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905/AnsiballZ_postgresql_ext.py
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> EXEC /bin/sh -c 'chmod u+x /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905/ /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905/AnsiballZ_postgresql_ext.py && sleep 0'
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> EXEC /bin/sh -c 'sudo -H -S -n  -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-itmgxinbjkymfglnajmwqruqtuheitlp ; /usr/bin/python3 /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905/AnsiballZ_postgresql_ext.py'"'"' && sleep 0'
<{{ famedly_instance_domain | regex_replace('\.(care|de)$', '.net') }}> EXEC /bin/sh -c 'rm -f -r /home/gitlab-runner/.ansible/tmp/ansible-tmp-1678973886.7891605-17637-250160229280905/ > /dev/null 2>&1 && sleep 0'
changed: [ci.famedly.local] => (item=workflow-engine-requeuest) => {
    "ansible_loop_var": "item",
    "changed": true,
    "db": "workflow-engine-requeuest",
    "ext": "uuid-ossp",
    "invocation": {
        "module_args": {
            "ca_cert": null,
            "cascade": false,
            "db": "workflow-engine-requeuest",
            "ext": "uuid-ossp",
            "login_host": "/opt/postgresql/sockets",
            "login_password": "",
            "login_unix_socket": "",
            "login_user": "postgres",
            "name": "uuid-ossp",
            "port": 5432,
            "schema": null,
            "session_role": null,
            "ssl_mode": "prefer",
            "state": "present",
            "trust_input": true,
            "version": null
        }
    },
    "item": "workflow-engine-requeuest",
    "queries": [
        "CREATE EXTENSION \"uuid-ossp\""
    ]
}
```
I don't really see anything helpful in there right now though.

**Comment by Andersson007** [comment]
@jcgruenhage thanks for sharing the output!
I've opened the issue and will put `help_wanted` in case anyone wants to investigate this

**Comment by Andersson007** [comment]
i guess it relates to
>"msg": "Management of PostgreSQL extension failed: duplicate key value violates unique constraint \"pg_extension_name_index\"\nDETAIL:  Key (extname)=(uuid-ossp) already exists.\n"

looks like uuid-ossp already exists and we try to add it again. the module should of course see this.
can there maybe another Ansible process creating the same object in parallel?

**Comment by jcgruenhage** [comment]
> can there be maybe another Ansible process creating the same object in parallel?

No, there is just one ansible process running there. I don't really have a clue what's going on here.

Maybe it'd make sense to add a few more log lines for `-vvv`, so that we can rule out it being a race condition with a second process?

**Comment by Andersson007** [comment]
> > can there be maybe another Ansible process creating the same object in parallel?
> 
> No, there is just one ansible process running there. I don't really have a clue what's going on here.
> 
> Maybe it'd make sense to add a few more log lines for `-vvv`, so that we can rule out it being a race condition with a second process?

we could try, though if you're sure only one process is running, not sure if we catch anything useful in the more verbose output.

**Comment by Andersson007** [comment]
let's wait anyone inspect the code (or i'll inspect when there's more free time)

**Comment by RealGreenDragon** [comment]
I checked the PostgreSQL source code (master branch) and seems I found the issue: its a data race.

See this comment in CreateExtension function:

https://github.com/postgres/postgres/blob/d088ba5a5aa410d39b64f013e8433ad9eb3d17f1/src/backend/commands/extension.c#L1758-L1763

Even though we check in postgresql_ext module if the extension already exists before creation, the unique index pg_extension.extname is the real protection against race conditions (when for example two backends try to create same extension at the same).

So seems it is a PostgreSQL expected behaviour.
